proposals/2022-09-14-kubernetes-versioning.md

```diff
@@ -0,0 +1,208 @@
+
+1. A shim binary is installed on a Kubernetes node at `/bin/containerd-shim-spin-v1`.
+2. The following shows the containerd config.toml that maps the runtime type to the shim binary from step 1.
+      ```toml
+        [plugins.cri.containerd.runtimes.spin]
+          runtime_type = "io.containerd.spin.v1"
+      ```
+3. A runtime class that contains a handler that matches the "spin" config runtime name from step 2.
+      ```yaml
+      apiVersion: node.k8s.io/v1
+      kind: RuntimeClass
+      metadata:
+        name: wasmtime-spin
+      handler: spin
+      ```
+    
+    **NOTE: if no scheduling configuration is specified, the Kubernetes scheduler expects every node is able to run the workload.**
+
+4. A Wasm workload with the specified runtime class name matching the "wasmtime-spin" runtime class from step 3.
+      ```yaml
+      apiVersion: apps/v1
+      kind: Deployment
+      metadata:
+        name: wasm-spin
+      spec:
+        replicas: 1
+        selector:
+          matchLabels:
+            app: wasm-spin
+        template:
+          metadata:
+            labels:
+              app: wasm-spin
+          spec:
+            runtimeClassName: wasmtime-spin
+            containers:
+            - name: spin-hello
+              image: ghcr.io/deislabs/containerd-wasm-shims/examples/spin-rust-hello:latest
+              command: ["/"]
+      ```
+
+The above configuration will work well if there are no breaking changes introduced to the shim binary or if no new features are added that a given workload depends upon. However, this scenario is unlikely given the speed of innovation in this space.
+
+To safe introduce new nodes and new versions of shims, we must have a versioning scheme for both the shim binary and the RuntimeClasses.
+
+### <a name='Goals'></a>Goals
+1) A user should be able to add new nodes to the cluster with new versions of the shims and not disturb existing workloads.
+2) A user should be able to target a major, major + minor, or major+minor+patch semantic version of a shim.
+
+### <a name='NonGoals'></a>Non Goals
+
+## <a name='Proposal'></a>Proposal
+Now that we have a better understanding of the problem, let's evaluate a solution proposal and possible alternatives.
+
+### <a name='VersioningofbothshimbinaryandRuntimeClassname'></a>Versioning of both shim binary and RuntimeClass name
+We will introduce both a version for the shim binary and a version in the RuntimeClass name. The subsequent structure should provide an illustration of the proposed solution.
+
+1. A shim binary is installed on a Kubernetes node at `/bin/containerd-shim-spin-v1.2.3-v1`. **Note:** the semantic version was added to the file name for the bin.
```

Comment:
Looking at this binary name `containerd-shim-spin-v1.2.3-v1`, I am wondering why the postfix `v1` version is still needed in this shim binary. It seems to me that it was introduced to reference `runwasi` version, which adopts a versioning scheme similar to `runc`. Since `runwasi` is a direct dependency of wasm shims, we no longer need this explicit version anymore, no?

Reply:
I think the v1 there indicates the version of the containerd shim plugin protocol.